core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala

```diff
@@ -491,7 +492,10 @@ private[deploy] class Worker(
       handleRegisterResponse(msg)
 
     case SendHeartbeat =>
-      if (connected) { sendToMaster(Heartbeat(workerId, self)) }
+      if (connected) {
+        sendToMaster(Heartbeat(workerId, self, leakedExecutorSinceLastHeartbeat))
```

Comment:
This would potentially include an extra HashSet if we have leaked executors, could we avoid this overhead?

Reply:
Like to detect the timeout executors from the driver side?

Reply:
> This would potentially include an extra HashSet if we have leaked executors, could we avoid this overhead?

It's only used for the corner case. So in most times, the set is empty.